### DISCUSSION TITLE
Leaflet map on modularized page

### DISCUSSION THREAD

**Discussion by Nikodemski2002** [zauberzeug/nicegui#554] [Q&A] [answered]
Hi everyone, I have a problem with getting the map example working but on a modularized page (in a separate .py file).
For simplicity, I will just post the code examples I've tried to get this working and their result in a descriptive and graphic form.
*Note: all pages also use auth as in the authentication example*

**Example 1 (await client.connected() inside main.py)**
<details>
  <summary>Example 1 details</summary>

**main.py code:**
```py
import uuid
from typing import Dict
from fastapi import Request
from nicegui import app, ui, Client
from fastapi.responses import RedirectResponse
from starlette.middleware.sessions import SessionMiddleware
from pages import map_page

@ui.page('/map')
async def map(request: Request, client: Client):
    if not is_authenticated(request):
        return RedirectResponse('/')
    session = session_info[request.session['id']]
    await client.connected()
    map_page.content(session)

ui.run()
```
**map_page.py code:**
```py
from leaflet import leaflet
from nicegui import ui
import nav_bar

locations = {
    (52.5200, 13.4049): 'Berlin',
    (40.7306, -74.0060): 'New York',
    (39.9042, 116.4074): 'Beijing',
    (35.6895, 139.6917): 'Tokyo',
}


def content(session):
    ui.colors(primary='#354B9B')
    nav_bar.nav_bar(session)
    ui.markdown('###**Map**').style('color: #4C4C4C; margin-right: auto; margin-left: auto')
    map = leaflet().classes('w-full h-96')
    selection = ui.select(locations, on_change=lambda e: map.set_location(e.value)).classes('w-40')
    selection.set_value(next(iter(locations)))  # trigger map.set_location with first location in selection
```
**Result:**
The page renders correctly including the `ui.select()` element, but the leaflet component doesn't show at all.
![image](https://user-images.githubusercontent.com/19575525/226142859-3b598913-b04f-4b20-81e7-ea60010825da.png)

</details>

---
**Example 2 (await client.connected() parsed to map_page.content())**
<details>
  <summary>Example 2 details</summary>

**main.py code:**
```py
import uuid
from typing import Dict
from fastapi import Request
from nicegui import app, ui, Client
from fastapi.responses import RedirectResponse
from starlette.middleware.sessions import SessionMiddleware
from pages import map_page

@ui.page('/map')
def map(request: Request, client: Client) -> None:
    if not is_authenticated(request):
        return RedirectResponse('/')
    session = session_info[request.session['id']]
    map_page.content(session, client)

ui.run()
```
**map_page.py code:**
```py
from leaflet import leaflet
from nicegui import ui
import nav_bar

locations = {
    (52.5200, 13.4049): 'Berlin',
    (40.7306, -74.0060): 'New York',
    (39.9042, 116.4074): 'Beijing',
    (35.6895, 139.6917): 'Tokyo',
}


async def content(session, client):
    ui.colors(primary='#354B9B')
    nav_bar.nav_bar(session)
    ui.markdown('###**Map**').style('color: #4C4C4C; margin-right: auto; margin-left: auto')
    map = leaflet().classes('w-full h-96')
    selection = ui.select(locations, on_change=lambda e: map.set_location(e.value)).classes('w-40')
    await client.connected()
    selection.set_value(next(iter(locations)))  # trigger map.set_location with first location in selection

```
**Result:**
The page doesn't render at all. The terminal returns RuntimeWarning: `coroutine 'content' was never awaited`

</details>

---
**Example 3 (await Client.connected() inside map_page.py)**
<details>
  <summary>Example 3 details</summary>

**main.py code:**
```py
import uuid
from typing import Dict
from fastapi import Request
from nicegui import app, ui
from fastapi.responses import RedirectResponse
from starlette.middleware.sessions import SessionMiddleware
from pages import map_page

@ui.page('/map')
def map(request: Request) -> None:
    if not is_authenticated(request):
        return RedirectResponse('/')
    session = session_info[request.session['id']]
    map_page.content(session)

ui.run()
```
**map_page.py code:**
```py
from leaflet import leaflet
from nicegui import ui, Client
import nav_bar

locations = {
    (52.5200, 13.4049): 'Berlin',
    (40.7306, -74.0060): 'New York',
    (39.9042, 116.4074): 'Beijing',
    (35.6895, 139.6917): 'Tokyo',
}


async def content(session):
    ui.colors(primary='#354B9B')
    nav_bar.nav_bar(session)
    ui.markdown('###**Map**').style('color: #4C4C4C; margin-right: auto; margin-left: auto')
    map = leaflet().classes('w-full h-96')
    selection = ui.select(locations, on_change=lambda e: map.set_location(e.value)).classes('w-40')
    await Client.connected()
    selection.set_value(next(iter(locations)))  # trigger map.set_location with first location in selection
```
**Result:**
The page doesn't render at all. The terminal returns RuntimeWarning: `coroutine 'content' was never awaited`

</details>

---
**Example 4 (ignore await client.connected())**
<details>
  <summary>Example 4 details</summary>

I've also tested it without using any await client.connected() but as expected it displayed the 'gray' map at the beginning, and the map required user interaction to show the location correctly.
![example-4](https://user-images.githubusercontent.com/19575525/226143812-88a5286d-492a-42e0-8029-0734af0d52f1.gif)
</details>

---
**Example 5 (use without modularization)**
In the end, I tried using it without any modularization (just in the main.py file), practically the same way as in the example just to check if the rest of the code was ok, and it worked 100% properly. 

Have a good day,
Nikodem

**Comment by falkoschindler** [comment]
That sound's like an interesting problem. Unfortunately I can't try out your code examples since there are things missing like `is_authenticated`, `session_info`, `nav_bar`, ... Most of them are probably not needed anyway. So can you, please, try to reduce the code to a [minimum reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)? Example 1 should be enough to reproduce the original problem. Thanks!

**Comment by rompgadgets** [comment]
Here is a simplified version of the bug as I was able to recreate it as well:

main.py
```py
from nicegui import Client, ui
import map_page

@ui.page('/')
async def index_page(client: Client) -> None:
    map_page.content()
    await client.connected()

ui.run()
```

map_page.py

```py
from leaflet import leaflet
from nicegui import ui

locations = {
    (52.5200, 13.4049): 'Berlin',
    (40.7306, -74.0060): 'New York',
    (39.9042, 116.4074): 'Beijing',
    (35.6895, 139.6917): 'Tokyo',
}

def content() -> None:
    ui.colors(primary='#354B9B')
    ui.markdown('###**Map**').style('color: #4C4C4C; margin-right: auto; margin-left: auto')
    map = leaflet().classes('w-full h-96')
    selection = ui.select(locations, on_change=lambda e: map.set_location(e.value)).classes('w-40')
    selection.set_value(next(iter(locations)))  # trigger map.set_location with first location in selection
```

Note: leaflet.js and leaflet.py are identical from the map example.

The map only actually renders the location through user interaction, the set_value call in the content of map_page doesn't seem to trigger the map to render.

**Reply by falkoschindler** [reply]
(Edit: I moved this comment over here to be able to mark it as the answer to this question: https://github.com/zauberzeug/nicegui/discussions/554#discussioncomment-5398155)

**Comment by falkoschindler** [comment, marked as answer]
Ok, now I see the problem: You need to await a client connection _before_ setting the location:

```py
@ui.page('/')
async def index_page(client: Client) -> None:
    map = leaflet().classes('w-full h-96')
    await client.connected()
    map.set_location((52.5200, 13.4049))
```

So you probably need to move the initial `selection.set_value` out of `content` and execute it _after_ `client.connected` has been awaited.
